src/ir/module-splitting.cpp

```diff
@@ -0,0 +1,405 @@
+        std::set<Name>& calledPrimaryFuncs;
+        CallCollector(const std::set<Name>& primaryFuncs,
+                      std::set<Name>& calledPrimaryFuncs)
+          : primaryFuncs(primaryFuncs), calledPrimaryFuncs(calledPrimaryFuncs) {
+        }
+
+        void visitCall(Call* curr) {
+          if (primaryFuncs.find(curr->target) != primaryFuncs.end()) {
+            calledPrimaryFuncs.insert(curr->target);
+          }
+        }
+      };
+      CallCollector(primaryFuncs, calledPrimaryFuncs).walkFunction(func);
+    });
+  std::set<Name> calledPrimaryFuncs;
+  for (auto& entry : callCollector.map) {
+    auto& calledFuncs = entry.second;
+    calledPrimaryFuncs.insert(calledFuncs.begin(), calledFuncs.end());
+  }
+
+  // Find exported primary functions and map to their export names
+  std::map<Name, Name> exportedPrimaryFuncs;
+  for (auto& ex : primary.exports) {
+    if (ex->kind == ExternalKind::Function) {
+      exportedPrimaryFuncs.insert(std::make_pair(ex->value, ex->name));
+    }
+  }
+
+  // Ensure each called primary function is exported and imported
+  for (auto primaryFunc : calledPrimaryFuncs) {
+    Name exportName;
+    auto exportIt = exportedPrimaryFuncs.find(primaryFunc);
+    if (exportIt != exportedPrimaryFuncs.end()) {
+      exportName = exportIt->second;
+    } else {
+      exportName = Names::getValidExportName(primary, primaryFunc);
+      primary.addExport(
+        new Export{exportName, primaryFunc, ExternalKind::Function});
+    }
+    auto func = std::make_unique<Function>();
+    func->module = importNamespace;
+    func->base = exportName;
+    func->name = primaryFunc;
+    func->sig = primary.getFunction(primaryFunc)->sig;
+    secondary.addFunction(std::move(func));
+  }
+}
+
+void setupTablePatching(Module& primary,
+                        Module& secondary,
+                        const std::set<Name>& secondaryFuncs,
+                        Name placeholderNamespace) {
+  std::map<Index, Name> replacedElems;
+  // Replace table references to secondary functions with an imported
+  // placeholder that encodes the table index in its name:
+  // `importNamespace`.`index`.
+  forEachElement(primary.table, [&](Index index, Name& elem) {
+    if (secondaryFuncs.find(elem) != secondaryFuncs.end()) {
+      replacedElems[index] = elem;
+      auto* secondaryFunc = secondary.getFunction(elem);
+      auto placeholder = std::make_unique<Function>();
+      placeholder->module = placeholderNamespace;
+      placeholder->base = std::to_string(index);
+      placeholder->name = Names::getValidFunctionName(
+        primary,
+        std::string("placeholder_") + std::string(placeholder->base.c_str()));
+      placeholder->hasExplicitName = false;
+      placeholder->sig = secondaryFunc->sig;
+      elem = placeholder->name;
+      primary.addFunction(std::move(placeholder));
+    }
+  });
+
+  // Create active table segments in the secondary module to patch in the
```

Comment:
Can table segments ever be passive?

Reply:
Yes, although that hasn't been implemented in Binaryen yet because no one is using it.